### PR TITLE
Solana block prices

### PR DESCRIPTION
**Pull request by couralex6**

### CODE REVIEW

**Comment by couralex6** [comment]
Should probably move this model to the solana subproject

**Comment by jeff-dude** [comment]
> Should probably move this model to the solana subproject

yes for sure

File: dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql [hunk 1]
```diff
@@ -0,0 +1,167 @@
+{{ config(
+    schema = 'dex'
```
Comment (jeff-dude) [inline dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql:2]:
```suggestion
    schema = 'dex_solana'
```

File: dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    , alias = 'prices_block_solana'
```
Comment (jeff-dude) [inline dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql:3]:
```suggestion
    , alias = 'prices_block'
```

File: dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql
Comment (jeff-dude) [inline dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql:1]: dbt_subprojects/dex/models/prices/dex_solana_prices_block.sql

File: dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    , incremental_predicates = [incremental_predicate('DBT_INTERNAL_DEST.block_time')]
+)
+}}
+
+with dex_trades_raw as (
+    select
+        blockchain
+        , block_number
+        , block_time
+        , token_bought_address
```
Comment (jeff-dude) [inline dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql:18]: column names differ a bit in `dex_solana.trades`, you will likely hit errors on this select

File: dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        , t.token_bought_amount_raw
+        , t.token_bought_amount
+        , t.token_sold_address
+        , t.token_sold_amount_raw
+        , t.token_sold_amount
+        , t.amount_usd
+    from
+        dex_trades_raw as t
+    --only output swaps which contain a trusted token
+    inner join {{ source('prices', 'trusted_tokens') }} as tt
```
Comment (jeff-dude) [inline dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql:50]: will want to validate we have solana tokens in here

File: dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        , t.symbol as symbol
+        , t.decimals as decimals
+        , d.block_number as block_number
+        , d.block_time as block_time
+        , d.token_bought_amount as amount
+        , d.amount_usd
+        , coalesce(d.amount_usd/d.token_bought_amount, d.amount_usd/(d.token_bought_amount_raw/POW(10, t.decimals))) as price
+    from
+        dex_trades as d
+    inner join {{ source('tokens', 'erc20') }} as t
```
Comment (jeff-dude) [inline dbt_subprojects/dex/models/prices/dex_prices_block_solana.sql:67]: not sure we have solana in this table

we will need to check in the solana lineage how it leverages decimals to calculate amount_usd and then do similar here

doubtful our automated source for erc20 has solana tokens

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Comment by jeff-dude** [comment]
until you get CI passing, i would suggest hardcoding a date filter for non-incremental to last few days to expedite

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        , array[token_bought_mint_address, token_sold_mint_address] as tokens_swapped
+    from
+        {{ ref('dex_solana_trades') }}
+    where
+        1 = 1
+        and amount_usd > 0
+        and block_time >= now() - interval '10' day
+        {# {% if is_incremental() %}
+        and {{ incremental_predicate('block_time') }}
+        {% endif %} #}
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:33]:
```suggestion
        {% if is_incremental() %}
        and {{ incremental_predicate('block_time') }}
        {% endif %}
```

Comment (jeff-dude) [reply dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:33]: lets keep this enabled along with the historical hardcoded filter

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        , t.token_bought_amount
+        , t.token_sold_mint_address
+        , t.token_sold_amount_raw
+        , t.token_sold_amount
+        , t.amount_usd
+    from
+        dex_trades_raw as t
+    --only output swaps which contain a trusted token
+    inner join {{ source('prices', 'trusted_tokens') }} as tt
+        on t.blockchain = 'solana'
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:50]:
```suggestion
        on t.blockchain = tt.blockchain
```

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        , d.block_time as block_time
+        , d.token_bought_amount as amount
+        , d.amount_usd
+        , coalesce(d.amount_usd/d.token_bought_amount, d.amount_usd/(d.token_bought_amount_raw/POW(10, t.decimals))) as price
+    from
+        dex_trades as d
+    inner join {{ ref('tokens_solana_fungible') }} as t
+        on d.token_bought_mint_address = t.token_mint_address
+    left join {{ source('prices', 'trusted_tokens') }} as ptt
+        on d.blockchain = 'solana'
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:68]:
```suggestion
        on d.blockchain = ptt.blockchain
```

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        , t.symbol as symbol
+        , t.decimals as decimals
+        , d.block_time as block_time
+        , d.token_sold_amount as amount
+        , d.amount_usd
+        , coalesce(d.amount_usd/d.token_sold_amount, d.amount_usd/(d.token_sold_amount_raw/POW(10, t.decimals))) as price
+    from
+        dex_trades as d
+    inner join {{ ref('tokens_solana_fungible') }} as t
+        on d.token_bought_mint_address = t.token_mint_address
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:91]:
```suggestion
        on d.token_sold_mint_address = t.token_mint_address
```

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    left join {{ source('prices', 'trusted_tokens') }} as ptt
+        on d.blockchain = 'solana'
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:93]:
```suggestion
        on d.blockchain = ptt.blockchain
```

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 11]
```diff
@@ -0,0 +1,159 @@
+{{ config(
+    schema = 'dex_solana'
+    , alias = 'prices_block'
+    , partition_by = ['block_month']
+    , materialized = 'incremental'
+    , file_format = 'delta'
+    , incremental_strategy = 'merge'
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:7]:
```suggestion
    , incremental_strategy = 'delete+insert'
```

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        , contract_address
+    having
+        sum(amount_usd) >= 10000
+)
+select
+    blockchain
+    , contract_address
+    , symbol
+    , decimals
+    , cast(date_trunc('month', block_time) as date) as block_month -- for partitioning
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:132]:
```suggestion
    , cast(date_trunc('day', block_time) as date) as block_date -- for partitioning
```

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 13]
```diff
@@ -0,0 +1,159 @@
+{{ config(
+    schema = 'dex_solana'
+    , alias = 'prices_block'
+    , partition_by = ['block_month']
```
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:4]:
```suggestion
    , partition_by = ['block_date']
```

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql [hunk 14]
```diff
@@ -0,0 +1,159 @@
+{{ config(
+    schema = 'dex_solana'
+    , alias = 'prices_block'
+    , partition_by = ['block_date']
+    , materialized = 'incremental'
+    , file_format = 'delta'
+    , incremental_strategy = 'delete+insert'
+    , unique_key = ['blockchain', 'contract_address', 'symbol', 'decimals', 'block_time']
+    , incremental_predicates = [incremental_predicate('DBT_INTERNAL_DEST.block_time')]
```
[the remark this answers is not in the document]
Comment (jeff-dude) [reply dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:9]: can remove this after moving away from merge strategy, that's only useful there

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql
Comment (jeff-dude) [inline dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:1]: dbt_subprojects/solana/models/_sector/dex/prices/dex_solana_prices_block.sql

Comment (jeff-dude) [reply dbt_subprojects/solana/models/_sector/dex/prices/dex_prices_block_solana.sql:1]: can you update file name to match table name?

**Review by jeff-dude — CHANGES_REQUESTED** [review]
in addition to file name below, can you put in schema.yml file similar to how evm version?

**Pull request merged**